Fixes #35084 - Add rerun to job wizard

MariaAga opened the pull request:
changed the menu url to `experimental/job_wizard/new`
to use this pr go to `experimental/job_wizard/_ID_/rerun` as its experimental.
UI changes are adding the org/location alerts.
Patternfly guidelines say `If an alert message applies globally to the content on a page, place the alert in the page header area just below the title.`
![Screenshot from 2022-06-20 11-59-18](https://user-images.githubusercontent.com/30431079/174616346-069fa058-536f-427b-a495-c2d46b7cce24.png)

MariaAga commented on app/controllers/ui_job_wizard_controller.rb:
```diff
@@ -57,4 +57,19 @@ def resources
+    job = JobInvocation.authorized.find(params[:id])
+    composer = JobInvocationComposer.from_job_invocation(job, params).params
+    job_template_inputs = JobTemplate.authorized.find(composer[:template_invocations][0][:template_id]).template_inputs_with_foreign
+    inputs = Hash[job_template_inputs.map{ |input| ["inputs[#{input[:name]}]", (composer[:template_invocations][0][:input_values].select {|value| value[:template_input_id]== input[:id]})[0][:value]] } ]
```
Is there a better way to do this? and does this function needs permissions anywhere?

adamruzicka commented on app/controllers/ui_job_wizard_controller.rb:
```diff
@@ -57,4 +57,19 @@ def resources
+  def job_invocation
+    job = JobInvocation.authorized.find(params[:id])
+    composer = JobInvocationComposer.from_job_invocation(job, params).params
+    job_template_inputs = JobTemplate.authorized.find(composer[:template_invocations][0][:template_id]).template_inputs_with_foreign
+    inputs = Hash[job_template_inputs.map{ |input| ["inputs[#{input[:name]}]", (composer[:template_invocations][0][:input_values].select {|value| value[:template_input_id]== input[:id]})[0][:value]] } ]
```
I'd do something like this. At least to me it reads a bit better.

```suggestion
    pattern_template_invocation = composer[:template_invocations].first
    job_template_inputs = JobTemplate.authorized.find(pattern_template_invocation[:template_id]).template_inputs_with_foreign
    inputs = job_template_inputs.reduce({}) do |acc, input|
      value = pattern_template_invocation[:input_values].find { |value| value[:template_input_id] == input[:id] }
      acc.merge("inputs[#{input[:name]}]" => value[:value])
    end

```

MariaAga replied:
To re-use the autofilling from the url which comes in this format. I can change it but I think since we already have a function that does the same, to re-use it. 
autofill pr for context: https://github.com/theforeman/foreman_remote_execution/pull/629

adamruzicka replied:
Ah, that makes sense. Let's keep it as is then

adamruzicka commented:
If I have where all the hosts succeeded and I do rerun failed only, in the old form the search would get pre-filled to `name ^ ()` and it would resolve to no hosts. In the wizard the search query becomes empty which then resolves to all hosts known to foreman.

MariaAga commented:
I can't find any code results for `rerun_failed`, where should it be passed and what is its value?

adamruzicka commented:
Ermm, the button is "rerun failed", the actual parameter is `failed_only`, sorry about the confusion. The value is either `true` or `1` or the parameter is not passed.

MariaAga commented:
Thanks for the review, added the `failed_only`

adamruzicka commented on app/controllers/ui_job_wizard_controller.rb:
```diff
@@ -57,4 +57,19 @@ def resources
+    job = JobInvocation.authorized.find(params[:id])
+    composer = JobInvocationComposer.from_job_invocation(job, params).params
+    job_template_inputs = JobTemplate.authorized.find(composer[:template_invocations][0][:template_id]).template_inputs_with_foreign
+    inputs = Hash[job_template_inputs.map { |input| ["inputs[#{input[:name]}]", (composer[:template_invocations][0][:input_values].select { |value| value[:template_input_id]== input[:id] })[0][:value]] } ]
```
Please do use `find` over `select {...}[0]`, there's no need to collect all the elements matching a predicate if we only need the first one

MariaAga commented on webpack/JobWizard/JobWizard.js:
```diff
@@ -48,7 +48,8 @@ export const JobWizard = ({ rerunData }) => {
   });
   const [hostsSearchQuery, setHostsSearchQuery] = useState('');
-  const [fills, setFills] = useState(useSelector(selectRouterSearch));
+  const routerSearch = useSelector(selectRouterSearch);
```
Can't put hooks inside conditions

MariaAga commented:
There was an issue with hooks timing for the filling, should work now

adamruzicka commented:
Waiting for someone to take a look at the js bits, as far as functionality goes +1 from me

jeremylenz commented on webpack/JobWizard/JobWizard.js:
```diff
@@ -102,23 +107,50 @@ export const JobWizard = () => {
+        ...rerunData.inputs,
+      });
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Can this effect be eliminated and instead calculate all the values during render?

https://beta.reactjs.org/learn/you-might-not-need-an-effect#how-to-remove-unnecessary-effects

jeremylenz replied:
If you feel strongly about keeping this an effect, at least the rule shouldn't be disabled.  But as I was saying before, I think the better solution is to not use an effect at all.

jeremylenz reviewed (changes requested):
What are the consequences to me as the user if I run the job from a different org / location?  Does this mean the job will run on different hosts than it did before?

Below, same comment as last time.  (sorry :grimacing:)

adamruzicka commented:
Needs a rebase now

MariaAga commented:
Thanks for the reviews!
Rebased, removed some logic from the useeffect, removed `eslint-disable`

> What are the consequences to me as the user if I run the job from a different org / location? Does this mean the job will run on different hosts than it did before?

From what I see the host count is different, @adamruzicka is there something else?

jeremylenz commented:
I'm getting 
```
No route matches [GET] "/experimental/job_wizard/493/rerun"
```
all of a sudden.  Not sure what I'm doing wrong..

MariaAga commented:
@jeremylenz I'm not seeing this issue, but when I had it I had to re-run the server

jeremylenz commented:
@MariaAga got it working now, thanks :)

adamruzicka commented:
The host resolution is the only thing that should be affected by org/loc selection

MariaAga commented:
Thanks! new message:
![Screenshot from 2022-07-13 13-54-52](https://user-images.githubusercontent.com/30431079/178727968-9f56dfea-3506-4660-9934-d091025c3295.png)

jeremylenz reviewed (approved):
Message changes look good :+1: 

Works well.

As I said in the previous review I think the useEffect refactor could still go further, but I'll leave that to the future.

adamruzicka reviewed (approved):
Works nicely, ACK

adamruzicka commented:
Thank you @MariaAga & @jeremylenz !
